daml/ContingentClaims/Internal/Claim.daml

```diff
@@ -0,0 +1,121 @@
+import DA.Foldable (Foldable(..))
+import DA.Text (unwords, intercalate)
+import DA.Traversable (Traversable(..))
+import Daml.Control.Recursion
+import Prelude hiding (fst, snd)
+
+-- | Used to model cashflows of instruments.
+-- See [quickstart](../QUICKSTART.md) for detailed explanation.
+-- In Petyon-Jones' paper, this is called 'Contract'.
+-- We renamed it to avoid ambiguity.
+-- * `t` and `x` respectively correspond to the `Observation`s input type and the resulting output type.
+-- * `a` is the representation of an asset, e.g. a `Text` ISIN code.
+-- * `o` is the representation of an observable, e.g. a `Text`.
+data Claim t x a o
+  = Zero
+      -- ^ Represents an absence of claims. Monoid `And` identity.
+  | One a
+      -- ^ The bearer acquires one unit of `a` *immediately*.
+  | Give (Claim t x a o)
+      -- ^ The obligations of the bearer and issuer are revesed.
+  | And with fst: Claim t x a o, snd: Claim t x a o, tail: [Claim t x a o]
+      -- ^ Used to combine multiple rights together.
+  | Or with fst: Claim t x a o, snd: Claim t x a o, tail: [Claim t x a o]
+      -- ^ Gives the bearer the right to choose between several claims.
+  | Cond with predicate: Inequality t x o, success: Claim t x a o, failure: Claim t x a o
+      -- ^ Gives the bearer the right to the first claim if `predicate` is true, else the second claim.
+  | Scale with k: Observation t x o, claim: Claim t x a o
+      -- ^ Multiplies the `claim` by `k` (which can be non-deterministic).
+  | When with predicate: Inequality t x o, claim: Claim t x a o
+      -- ^ Defers the acquisition of `claim` until *the first instant* that `predicate` is true.
+  | Anytime with predicate: Inequality t x o, claim: Claim t x a o
+      -- ^ Like `When`, but valid any time the predicate is true (not just infinium).
+  | Until with predicate: Inequality t x o, claim: Claim t x a o
+      -- ^ Expires said claim on the *first instant* that `predicate` is true.
+  deriving (Eq)
+
+-- | Unfixed version of `Claim`, for use with `Daml.Control.Recursion`.
+data ClaimF t x a o b
+  = ZeroF
+  | OneF a
+  | GiveF b
+  | AndF with fst: b, snd: b, tail : [b]
+  | OrF with fst: b, snd: b, tail : [b]
+  | CondF with predicate: Inequality t x o, success: b, failure: b
+  | ScaleF with k: Observation t x o, claim: b
+  | WhenF with predicate: Inequality t x o, claim: b
+  | AnytimeF with predicate: Inequality t x o, claim: b
+  | UntilF with predicate: Inequality t x o, claim: b
+  deriving (Functor)
+
+instance Recursive (Claim t x a o) (ClaimF t x a o) where
+  project Zero = ZeroF
+  project (One a) = OneF a
+  project (Give c) = GiveF c
+  project (And c1 c2 cs) = AndF c1 c2 cs
+  project (Or c1 c2 cs) = (OrF c1 c2 cs)
+  project (Cond p c c') = CondF p c c'
+  project (Scale k c) = ScaleF k c
+  project (When p c) = WhenF p c
+  project (Anytime p c) = AnytimeF p c
+  project (Until p c) = UntilF p c
+
+instance Corecursive (Claim t x a o) (ClaimF t x a o) where
```

Comment:
I'm not 100% sure if typeclass instances need to be in (non-internal) `Claim` so that they are automatically exported. Do you know?

Reply:
Could you check this? The idea is that doing import `CC.Claim._` should include these instances. Otherwise, you would need to also import `CC.Internal.Claim._`, which defeats the purpose of the exercise.

Reply:
I just checked and in order to use e.g. `project` one does not need to import `CC.Internal.Claim` so we are good to go